controller/src/lib.rs

```diff
@@ -0,0 +1,51 @@
+use anyhow::Result;
+use chrono::Utc;
+use dis_spawner::{
+    messages::agent::DroneStatusMessage,
+    messages::scheduler::{ScheduleRequest, ScheduleResponse},
+    nats::TypedNats,
+};
+use scheduler::Scheduler;
+use tokio::select;
+use tokio_stream::StreamExt;
+
+mod scheduler;
+
+pub async fn run_scheduler(nats: TypedNats) -> Result<()> {
+    let scheduler = Scheduler::default();
+    let mut spawn_request_sub = nats.subscribe(ScheduleRequest::subscribe_subject()).await?;
+
+    let mut status_sub = Box::pin(
+        nats.subscribe_jetstream(
+            &DroneStatusMessage::stream_name(),
+            &DroneStatusMessage::subscribe_subject(),
+        )
+        .await,
+    );
+
+    loop {
+        select! {
+            status_msg = status_sub.next() => {
+                let status_msg = status_msg.unwrap();
+
+                scheduler.update_status(Utc::now(), &status_msg);
+            },
+
+            spawn_request = spawn_request_sub.next() => {
+                let spawn_request = spawn_request.unwrap().unwrap();
+
+                let result = match scheduler.schedule() {
+                    Ok(drone_id) => {
+                        match nats.request(&spawn_request.value.schedule(drone_id)).await {
+                            Ok(false) | Err(_) => ScheduleResponse::NoDroneAvailable,
```

Comment:
Do we want to differentiate between NoDroneAvailable and other types of errors?

Reply:
My thinking is that `NoDroneAvailable` will be a user-exposed error, where everything else is an internal error. The justification is that `NoDroneAvailable` could be the fault of the user who runs the cluster, whereas other errors are our fault.